firedrake/slate/slac/utils.py

```diff
@@ -136,6 +152,113 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+def slate_to_gem(expression):
+    """ Method encapsulating stage 1 of tsslac. """
+    mapper, var2terminal = slate2gem(expression)
+    if not (type(mapper) == Indexed or type(mapper) == FlexiblyIndexed):
+        shape = mapper.shape if len(mapper.shape) != 0 else (1,)
+        mapper = Indexed(mapper, make_indices(len(shape)))
+    return list([mapper]), var2terminal
+
+
+@singledispatch
+def _slate2gem(expr, self):
+    raise AssertionError("Cannot handle terminal type: %s" % type(expr))
+
+
+@_slate2gem.register(sl.Tensor)
+@_slate2gem.register(sl.AssembledVector)
+def _slate2gem_tensor(expr, self):
+    shape = expr.shape if not len(expr.shape) == 0 else (1, )
+    name = f"T{len(self.var2terminal)}"
+    assert expr not in self.var2terminal.values()
+    var = Variable(name, shape)
+    self.var2terminal[var] = expr
+    return var
+
+
+@_slate2gem.register(sl.Block)
+def _slate2gem_block(expr, self):
+    child, = map(self, expr.children)
+
+    # get index of first block in a range of blocks
+    first_ind = ()
+    is_range = lambda index: (type(index) == range)
+    for index in expr._indices:
+        first_ind += (index.start,) if is_range(index) else (index,)
+
+    # get offset of the first index
+    offset = ()
+    for i, idx in enumerate([idx for idx in first_ind]):
+        offset += ((sum(expr.children[0].shapes[i][:idx])), )
+
+    extent = tuple(sum(shape) for shape in expr.shapes.values())
+
+    # generate a FlexiblyIndexed (sliced view)
+    slices = []
+    for idx, extent in list(zip(offset, extent)):
+        slices.append(slice(idx, idx+extent))
+    v = view(child, *slices)
```

Comment:
I'd prefer, although this is style, the haskellish

```
return view(child, *(slice(idx, idx+extent) for idx, extent in zip(offset, extent)))
```

A general note, no need to turn a generator into a list before iterating over it.